src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs

```diff
@@ -26,16 +26,36 @@ public static SafeMsQuicConfigurationHandle Create(QuicClientConnectionOptions o
 
             if (options.ClientAuthenticationOptions != null)
             {
+                SslClientAuthenticationOptions clientAuthenticationOptions = options.ClientAuthenticationOptions;
+
 #pragma warning disable SYSLIB0040 // NoEncryption and AllowNoEncryption are obsolete
-                if (options.ClientAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.NoEncryption)
+                if (clientAuthenticationOptions.EncryptionPolicy == EncryptionPolicy.NoEncryption)
                 {
-                    throw new PlatformNotSupportedException(SR.Format(SR.net_quic_ssl_option, nameof(options.ClientAuthenticationOptions.EncryptionPolicy)));
+                    throw new PlatformNotSupportedException(SR.Format(SR.net_quic_ssl_option, nameof(clientAuthenticationOptions.EncryptionPolicy)));
                 }
 #pragma warning restore SYSLIB0040
 
-                if (options.ClientAuthenticationOptions.ClientCertificates != null)
+                if (clientAuthenticationOptions.LocalCertificateSelectionCallback != null)
+                {
+                    X509Certificate? cert = clientAuthenticationOptions.LocalCertificateSelectionCallback(
+                        options,
+                        clientAuthenticationOptions.TargetHost ?? string.Empty,
+                        clientAuthenticationOptions.ClientCertificates ?? new X509CertificateCollection(),
+                        null,
+                        Array.Empty<string>());
+
+                    try
+                    {
+                        if (((X509Certificate2)cert).HasPrivateKey)
+                        {
+                            certificate = cert;
+                        }
+                    }
+                    catch { }
```

Comment:
Why might throw here, and do we need to eat all exceptions or just specific ones?

Also, `cert` above is nullable, but we're not checking for null?

Reply:
for server, we run this on msquic thread AFAIK. However, for client this is called from `MsQuicConnection` constructor. So we can perhaps surface any exception from the callback. (including the cast and perhaps check for PrivateKey.

Reply:
BTW it should be OK for the callback return NULL. It is up to the server to decide if no certificate is OK or not. I think SslStram has some tests for that case.

Reply:
I see that the only way this could throw is if it is not `X509Certificate2` instance or it was `.Reset()`ed, I will remove the try-catch and replace it with explicit checks.